feat(cart): remove coupon

nsilva-spryker opened the pull request:
- add remove button to delete coupon in coupon component
- update services and adapters to remove coupon and update cart totals
- update component unit and e2e tests coverage

closes: [HRZ-90814](https://spryker.atlassian.net/browse/HRZ-90814)


[HRZ-90814]: https://spryker.atlassian.net/browse/HRZ-90814?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

nsilva-spryker committed:
feat: implement remove coupon (#1059)

- add remove icon to delete coupon
- update services and adapters to remove coupon and update cart totals
- update component unit tests coverage

closes: [HRZ-91012](https://spryker.atlassian.net/browse/HRZ-91012)


[HRZ-91012]:
https://spryker.atlassian.net/browse/HRZ-91012?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

---------

Co-authored-by: tobi-or-not-tobi <tobi-or-not-tobi@users.noreply.github.com>

nsilva-spryker committed:
Merge branch 'development' into feature/hrz-90814-remove-coupon

Sinucid commented on libs/domain/cart/coupon/coupon.component.ts:
```diff
@@ -67,7 +67,17 @@ export class CouponComponent extends CartComponentMixin(
+              <span class="code">
+                ${coupon.code}
+                <oryx-button
+                  .type=${ButtonType.Icon}
```
need to add aria-label, the button without text inside does not pass accessibility check

Sinucid commented on libs/domain/cart/services/src/adapter/default-cart.adapter.ts:
```diff
@@ -163,6 +163,22 @@ export class DefaultCartAdapter implements CartAdapter {
+    return this.identity.get().pipe(
+      take(1),
+      switchMap((identity) => {
+        const url = this.generateUrl(
+          identity.isAuthenticated
+            ? `${ApiCartModel.UrlParts.Carts}/${data.cartId}/${ApiCartModel.UrlParts.Coupons}/${data.code}`
+            : `${ApiCartModel.UrlParts.GuestCarts}/${data.cartId}/${ApiCartModel.UrlParts.Coupons}/${data.code}`,
+          !identity.isAuthenticated
+        );
+
+        return this.http.delete(url);
```
```suggestion
        const requestType =  identity.isAuthenticated ? 
        ApiCartModel.UrlParts.Carts: 
        ApiCartModel.UrlParts.GuestCarts;
        
        const url = this.generateUrl(          
          `${requestType}/${data.cartId}/${ApiCartModel.UrlParts.Coupons}/${data.code}`,
          !identity.isAuthenticated
        );

        return this.http.delete(url);
```

Sinucid commented on libs/domain/cart/services/src/default-cart.service.ts:
```diff
@@ -101,12 +102,20 @@ export class DefaultCartService implements CartService {
+    action: (qualifier: CouponQualifier) => {
+      return this.adapter.deleteCoupon(qualifier);
+    },
+    onSuccess: [...this.cartCommandBase.onSuccess, CartEntryRemoved],
```
Probably query events are mixed up, this uses `CartEntryRemoved`, and entry command - `CouponRemoved`

Sinucid commented on libs/domain/cart/services/src/default-cart.service.ts:
```diff
@@ -287,6 +296,10 @@ export class DefaultCartService implements CartService {
     return this.executeWithOptionalCart(qualifier, this.addCouponCommand$);
   }
 
+  deleteCoupon(qualifier: CouponQualifier): Observable<unknown> {
```
and need to add this method to the cart mock

nsilva-spryker replied:
Coupon service will be set in different service and adapter in a follow up story, pls check: https://spryker.slack.com/archives/C029XPSBXC6/p1704820725966759

nsilva-spryker committed:
feat: refactor cr

nsilva-spryker committed:
feat: apply e2e test to remove coupon (#1067)

Update e2e tests when removing coupon

closes: [HRZ-90816](https://spryker.atlassian.net/browse/HRZ-90816)



[HRZ-90816]:
https://spryker.atlassian.net/browse/HRZ-90816?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
